File: app/src/main/java/ep/asynctask/MainActivity.java
package ep.asynctask;

import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getCanonicalName();

    private Button button;
    private TextView tv;
    private ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.i(TAG, "Metode GUI poganja nit " + Thread.currentThread());

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        button = (Button) findViewById(R.id.btnStart);
        tv = (TextView) findViewById(R.id.tvProgress);
        progressBar = (ProgressBar) findViewById(R.id.progressBar);
        progressBar.setMax(100);

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final MyAsyncTask task = new MyAsyncTask();
                task.execute(5);
            }
        });
    }

    private class MyAsyncTask extends AsyncTask<Integer, Double, Void> {
        @Override
        protected Void doInBackground(Integer... params) {
            for (int i = 0; i < params[0]; i++) {
                simulateWork();
            }

            return null;
        }

        private void simulateWork() {
            Log.i(TAG, "Simuliramo opravilo ... čakamo 1 sekundo. Nit = " + Thread.currentThread());
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Log.e(TAG, "Interrupted", e);
            }
        }
    }
}

File: app/src/main/java/fri/ep/MainActivity.kt
package fri.ep

import android.os.AsyncTask
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    private val tag = MainActivity::class.java.canonicalName
    var task: MyAsyncTask? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        progressBar.max = 100

        btnStart.setOnClickListener {
            task = MyAsyncTask(this)
            task?.execute(50)
        }

        Log.i(tag, "Metode GUI poganja nit " + Thread.currentThread())
    }

    override fun onStop() {
        task?.cancel(true)
        task = null
        super.onStop()
    }


    class MyAsyncTask(val activity: MainActivity) : AsyncTask<Int, Double, Unit>() { //Int -> vhodni, Double -> napredek, Unit -> rezultat
        val tag = MyAsyncTask::class.java.canonicalName

        override fun doInBackground(vararg params: Int?): Unit {
            for (i in 0 until params[0]!!) {
                if (isCancelled) {
                    break
                }

                simulateWork()
                publishProgress(i.toDouble() / params[0]!!.toDouble() * 100)
            }
        }

        override fun onProgressUpdate(vararg values: Double?) {
            activity.tvProgress.text = String.format("%.2f %%", values[0])
            activity.progressBar.progress = values[0]!!.toInt()
            // Log.i(tag, "Sporocamo napredek (${values[0]}). Nit = ${Thread.currentThread()}")

        }

        override fun onPostExecute(result: Unit?) {
            activity.tvProgress.text = "Completed"
            activity.progressBar.progress = 100
        }

        private fun simulateWork() {
            Log.i(tag, "Simuliramo opravilo ... čakamo 1 sekundo. Nit = ${Thread.currentThread()}")
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                Log.e(tag, "Interrupted", e)
            }
        }
    }

}
